File: 2-uninitialized-pointer-dereference/code.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct {
    double length;
    double width;
    double* area;     
    double* perimeter; 
} Rectangle;

void calculateArea(Rectangle* r) {
    // Allocate memory for area, but forget to initialize
    r->area = (double*)malloc(sizeof(double));
    *(r->area) = r->length * r->width;
}

void printInfo(Rectangle* r) {
    printf("Rectangle: %.2f x %.2f\n", r->length, r->width);
    printf("Area: %.2f\n", *(r->area));    // Safe if calculateArea was called
    printf("Perimeter: %.2f\n", *(r->perimeter)); // Uninitialized pointer dereference
}

Rectangle* createRectangle(double length, double width) {
    Rectangle* r = (Rectangle*)malloc(sizeof(Rectangle));
    r->length = length;
    r->width = width;
    // r->area and r->perimeter are not initialized
    return r;
}

void freeRectangle(Rectangle* r) {
    if (r->area != NULL) {
        free(r->area);
    }
    // Not checking r->perimeter because it might not be initialized
    // This can cause problems if freeing an uninitialized pointer
    if (r->perimeter != NULL) {  // Potential uninitialized pointer dereference
        free(r->perimeter);
    }
    free(r);
}

int main() {
    clock_t start, end;
    double cpu_time_used;
    
    start = clock();
    
    Rectangle* rect1 = createRectangle(5.0, 3.0);
    Rectangle* rect2 = createRectangle(7.5, 2.5);
    
    calculateArea(rect1);
    
    printf("Rectangle 1 Information:\n");
    printInfo(rect1);  // Uninitialized pointer dereference for perimeter
    
    printf("\nRectangle 2 Information:\n");
    printInfo(rect2);  // Uninitialized pointer dereference for both area and perimeter
    
    // Free memory
    freeRectangle(rect1);
    freeRectangle(rect2);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("\nExecution time: %f seconds\n", cpu_time_used);
    
    return 0;
}

File: 2-uninitialized-pointer-dereference/code.rs
use std::time::Instant;

struct Rectangle {
    length: f64,
    width: f64,
    area: *mut f64,     // Pointer to area calculation result
    perimeter: *mut f64, // Pointer to perimeter calculation result
}

fn calculate_area(r: &mut Rectangle) {
    unsafe {
        r.area = Box::into_raw(Box::new(0.0));
        *(r.area) = r.length * r.width;
    }
}

fn print_info(r: &Rectangle) {
    println!("Rectangle: {:.2} x {:.2}", r.length, r.width);
    
    // Dereference area pointer (potentially uninitialized)
    unsafe {
        println!("Area: {:.2}", *(r.area));
    }
    
    // Dereference perimeter pointer (uninitialized)
    unsafe {
        println!("Perimeter: {:.2}", *(r.perimeter)); // Uninitialized pointer dereference!
    }
}

fn create_rectangle(length: f64, width: f64) -> Rectangle {
    Rectangle {
        length,
        width,
        area: std::ptr::null_mut(),      // null pointer in Rust
        perimeter: std::ptr::null_mut(), // null pointer in Rust
    }
}

fn free_rectangle(r: Rectangle) {
    unsafe {
        if !r.area.is_null() {
            drop(Box::from_raw(r.area));
        }
        
        if !r.perimeter.is_null() {
            drop(Box::from_raw(r.perimeter));
        }
        
        // Struct r will be freed automatically when it goes out of scope
    }
}

fn main() {
    let start = Instant::now();
    
    let mut rect1 = create_rectangle(5.0, 3.0);
    let rect2 = create_rectangle(7.5, 2.5);
    
    calculate_area(&mut rect1);
    
    // Print information (will dereference uninitialized pointers)
    println!("Rectangle 1 Information:");
    print_info(&rect1);  // Uninitialized pointer dereference for perimeter
    
    println!("\nRectangle 2 Information:");
    print_info(&rect2);  // Uninitialized pointer dereference for both area and perimeter
    
    free_rectangle(rect1);
    free_rectangle(rect2);
    
    let duration = start.elapsed();
    println!("\nExecution time: {:?}", duration);
}